upb/def.c

```diff
@@ -1018,14 +1018,20 @@ static int field_number_cmp(const void *p1, const void *p2) {
   return f1->number - f2->number;
 }
 
-static void assign_layout_indices(const upb_msgdef *m, upb_msglayout_field *fields) {
+static void assign_layout_indices(const upb_msgdef *m, upb_msglayout *l,
+                                  upb_msglayout_field *fields) {
   int i;
   int n = upb_msgdef_numfields(m);
+  int dense_below = 0;
   for (i = 0; i < n; i++) {
     upb_fielddef *f = (upb_fielddef*)upb_msgdef_itof(m, fields[i].number);
     UPB_ASSERT(f);
     f->layout_index = i;
+    if (upb_fielddef_number(f) == i + 1) {
+      dense_below = upb_fielddef_number(f);
+    }
   }
+  l->dense_below = dense_below;
```

Comment:
As a potential heuristic:

- once I have k gaps within N of each other, I will rollback to the last gap and stop being dense

so if you say 3 gaps within 8 of each other, your worst case loss to gaps is 1/4.

Reply:
Yes seems possibly promising. Though some users may wish to always prioritize code size. Maybe less size-conscious users are better served with fast tables? It seems like promising future work.

Reply:
agreed, future work